active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+
+Engines provide an alternative to these approaches that allows for distributed
+development, testing, and packaging, _as well as_ logical run-time separation.
+Because engines are derived from applications, they can be just as
+full-featured. Each has its own namespace and registry. Even though engines are
+isolated from the applications that contain them, the boundaries between them
+allow for controlled sharing of resources.
+
+Engines can be either "routable" or "route-less":
+
+* Routable engines provide a routing map which can be integrated with the
+  routing maps of parent applications or engines. Routing maps are alway eager
+  loaded, which allows for deep linking into an engine's routes regardless of
+  whether the engine itself has been instantiated.
+
+* Route-less engines can isolate complex functionality that is not related to
+  routing (e.g. a chat engine in a sidebar). Route-less engines can be rendered
+  into outlets ad hoc as routes are loaded.
+
+The potential scope of engines is large enough that this feature merits
+development and delivery in multiple phases. A minimum viable version could be
+released sooner, which could be augmented with more advanced features later.
+
+An initial release of engines could provide the following benefits:
+
+* Distributed development - Engines can be developed and tested in isolation
+  within their own Ember CLI projects and included by applications or other
+  engines. Engines can be packaged and released as addons themselves.
+
+* Integrated routing - Support for mounting routable engines in the routing maps
+  of applications or other engines.
+
+* Ad hoc embedding - Support for embedding route-less engines in outlets as
+  needed.
+
+* Clean boundaries - An engine can cooperate with its parents through a few
+  explicit interfaces. Beyond these interfaces, engines and applications are
+  isolated.
+
+Subsequent releases of engines could allow for the following:
+
+* Lazy loading - An engine could allow its parent to boot with only its routing
+  map loaded. The rest of the engine could be loaded only as required (i.e.
+  when a route in an engine is visited). This would allow applications to boot
+  faster and limit their memory consumption.
+
+* Namespaced access to engine resources from applications - This could open up
+  the potential for applications to use, and extend, an engine's resources much
+  like resources in other addons, but without the possibility of namespace
+  collisions.
+
+## Detailed design
+
+Engines are very similar to regular applications: they can be developed in
+isolation in Ember CLI, include addons, and contain all the same elements,
+including routes, components, initializers, etc. The primary differences are
+that an engine does not boot itself and an engine does not control the router.
+
+### Engine internals
+
+New `Engine` and `EngineInstance` classes will be introduced.
+
+Applications and engines will share ancestry. It remains TBD whether
```

Comment:
i believe Applications must behave like or very similar to engines, we should be careful (and not make the mistake rails did here). The ambiguity in rails, causes much grief. E.G not namespacing apps, resulted in collisions between model names and engine modules...